File: 回溯/22.括号生成.java
class Solution {
    List<String> res;
    List<Character> path;
    public List<String> generateParenthesis(int n) {
        res=new ArrayList<>();
        path = new ArrayList<>();
        dfs(0,0,n);
        return res;
    }

    void dfs(int left,int right,int n){
        if(right==n){
            StringBuilder sb = new StringBuilder();
            for(char ch:path){
                sb.append(ch);
            }
            res.add(sb.toString());
            return;
        }
        if(left<n){
            path.add('(');
            left++;
            dfs(left,right,n);
            left--;
            path.removeLast();
        }
        if(right<left){
            path.add(')');
            right++;
            dfs(left,right,n);
            right--;
            path.removeLast();
        }
    }
}

File: 回溯/22.括号生成.go
func generateParenthesis(n int) []string {
	res := make([]string, 0)
	path := []byte{}
	cnt1, cnt2 := 0, 0
	var dfs func()
	dfs = func() {
		if cnt2 == n {
			res = append(res, string(path))
			return
		}
		if cnt1 < n {
			path = append(path, '(')
			cnt1++
			dfs()
			cnt1--
			path = path[:len(path)-1]
		}
		if cnt2 < cnt1 {
			path = append(path, ')')
			cnt2++
			dfs()
			cnt2--
			path = path[:len(path)-1]
		}
	}
	dfs()
	return res
}